### PR TITLE
Make CoreFX/Native be valid C++.

### PR DESCRIPTION
**Pull request by jaykrell**
- Assignment from void* to non-void.
    Add cast.
 - sizeof_member(struct foo, bar)
   Add typedef struct foo foo and remove struct in member_sizeof.

### CODE REVIEW

**Comment by danmosemsft** [comment]
Just curious what the value is of making it valid C++? I don't object..

**Comment by jaykrell** [comment]
To make mono C++, which are considering.
First would like it to build as C++ and verify it still works.
See https://github.com/mono/mono/pull/9722

File: src/Native/Unix/System.Native/pal_networking.c [hunk 1]
```diff
@@ -536,7 +544,7 @@ int32_t SystemNative_SetAddressFamily(uint8_t* socketAddress, int32_t socketAddr
 {
     struct sockaddr* sockAddr = (struct sockaddr*)socketAddress;
     if (sockAddr == NULL || socketAddressLen < 0 ||
-        !IsInBounds((const uint8_t*)sockAddr, (size_t)socketAddressLen, (const uint8_t*)&sockAddr->sa_family, sizeof_member(struct sockaddr, sa_family)))
+        !IsInBounds((const uint8_t*)sockAddr, (size_t)socketAddressLen, (const uint8_t*)&sockAddr->sa_family, sizeof_member(sockaddr, sa_family)))
```
Comment (jaykrell) [inline src/Native/Unix/System.Native/pal_networking.c]: Suggest IsInBounds take a const void* and then remove all the casts on inputs to it.

**Comment by filipnavara** [comment]
Eh, Mono just went exactly the opposite way - by converting the PAL from C++ to C and upstreaming that, which was only recently completed.

**Comment by jaykrell** [comment]
That is the current state but we are considering C++.
Anyway, I don't have to compile CoreFX as C++, just need the headers to be valid, and that was already fixed.

**Comment by danmosemsft** [comment]
@marek-safar ?

**Comment by jaykrell** [comment]
I can workaround and compile just CoreFX/native as C.
I discovered that later.
i.e. https://github.com/mono/mono/pull/9731/commits/364972d9d2ddc2f9fbe1b3d506811186437880d9#diff-65ca3a126a25b819a824e0397c36650aR127

**Comment by jaykrell** [comment]
These trivial changes alleviate messiness in mono/metadata/Makefile.am.

**Comment by am11** [comment]
@jaykrell, I couldn't find the compiler error you have received in the discussion thread (could you share?), but I think the problem is the way `-xc++` flag being set in your config.ac, that forces the compiler to compile all code files (including those with `.c` extension) as C++ code. C and C++ are different languages, so making `.c` file compile as C++ language is probably not the best way and will present more challenges in future. Your config.ac file defines the new option as:

`AC_ARG_ENABLE(cplusplus, [  --enable-cplusplus   compile some code as C++]) `

it seems like later it is missing the filter / glob for `*.c` vs. `*.cpp|*.cxx` to abide by the option description `compile *some* code as C++`, instead of `compile *all* code as C++`.

Out of interest, could you please share some information what is the overarching purpose of the `--cpluscplus` option being implemented?

**Comment by jaykrell** [comment]
@am11 
I'm aware of what I am doing wrt -xc++ and such and what the two languages are.
C is almost but not quite a proper subset of C++.
You can change your C code "just a little", for some definition of "little", to make it valid C++. And the consider going more into C++. I had it working back in January/February but it took a hiatus and now I have to merge with  master, and CoreFX/Native has come in since then.

The goal is:
1. Decide if we want to use C++.
2. As part of this decision, get the tree compiling as both C and C++ with minimum change.
i.e find the common subset, which is almost all of C. I'm far along that work.
At this point, or maybe forever, I do not want to rename any files. Renaming all the files would be too much change while we are deciding, and while we are fixing them to be both valid C and valid C++.
I also don't like renaming due to the ability to diff two directory  trees from before/after the rename -- no matter git's abilities.
3. Test the C++ version.  Verify its size, dependencies, correctness.
4. Decide.

What I did miss at first though, is that mono/metadata/Makefile.am builds a few libraries, and we can apply -xc++ to all of them  except for CoreFX/Native. That should help much.
We might still need header changes, but I'll see later.

But these are also small innocuous changes. I am not interested in fancier C++ here in CoreFX/Native, just a valid subset of C and C++.

**Comment by marek-safar** [comment]
I don't see any real value in compiling PAL with C++ but I would not oppose C++ compatible headers

**Comment by filipnavara** [comment]
The headers are already C++ compatible AFAIK.

**Comment by danmosemsft** [comment]
@jkotas

File: src/Native/Unix/System.Native/pal_networking.c (hunk 1 above)
Comment (jkotas) [reply src/Native/Unix/System.Native/pal_networking.c]: Yes, that would be nice.

File: src/Native/Unix/System.Native/pal_networking.c [hunk 2]
```diff
@@ -142,11 +142,19 @@ c_static_assert(GetHostErrorCodes_NO_DATA == NO_DATA);
...
+typedef struct iovec iovec;
+typedef struct sockaddr sockaddr;
+typedef struct xsocket xsocket;
+typedef struct linger linger;
+
 // We require that IOVector have the same layout as iovec.
 c_static_assert(sizeof(IOVector) == sizeof(struct iovec));
-c_static_assert(sizeof_member(IOVector, Base) == sizeof_member(struct iovec, iov_base));
+c_static_assert(sizeof_member(IOVector, Base) == sizeof_member(iovec, iov_base));
 c_static_assert(offsetof(IOVector, Base) == offsetof(struct iovec, iov_base));
```
Comment (jkotas) [inline src/Native/Unix/System.Native/pal_networking.c]: It would be nice to drop `struct` for offsetof here as well to make this look symmetric.

**Comment by jkotas** [comment]
> I don't see any real value in compiling PAL with C++ but I would not oppose C++ compatible headers

Are there changes required in the headers to make them C++ compatible?

**Comment by jkotas** [comment]
> these are also small innocuous changes

Agree (and the explicit casts for malloc are improvement). I think it is ok to take these if this helps something.

File: src/Native/Unix/System.Native/pal_networking.c (hunk 2 above)
Comment (jaykrell) [reply src/Native/Unix/System.Native/pal_networking.c]: Agreed, I just hadn’t checked for sufficiently typedefs, which appear to be there, or if not would be trivial. Will do.

File: src/Native/Unix/System.Native/pal_networking.c (hunk 1 above)
Comment (jaykrell) [reply src/Native/Unix/System.Native/pal_networking.c]: Will do separately later thank you.

**Comment by jaykrell** [comment]
The headers are probably already ok.
The error was something to do with sizeof(struct foo::bar) requiring bar to be a struct type within foo, not a member data. "struct implies type".

File: src/Native/Unix/System.Native/pal_networking.c [hunk 3]
```diff
@@ -438,8 +446,11 @@ int32_t SystemNative_GetHostName(uint8_t* name, int32_t nameLength)
     return gethostname((char*)name, unsignedSize);
 }
 
-static bool IsInBounds(const uint8_t* baseAddr, size_t len, const uint8_t* valueAddr, size_t valueSize)
+static bool IsInBounds(const void* void_baseAddr, size_t len, const void* value_valueAddr, size_t valueSize)
```
Comment (jkotas) [inline src/Native/Unix/System.Native/pal_networking.c]: `value_valueAddr` -> `void_valueAddr`

**Pull request merged**